### ISSUE TITLE
Crash with BETWEEN inside CASE expression

### ISSUE THREAD

**Issue by gladykov** [sql-formatter-org/sql-formatter#500]
**Describe the bug**

```
with
  p as (
    SELECT distinct
      s.id,
      case
        when s.end_date between (
          CASE
            WHEN CURRENT_DATE >= 12 THEN 13
          end
        ) and (
          CASE
            WHEN CURRENT_DATE >= 12 THEN 13
          end
        )  then year(current_date) + 1

      end as s_u,
    FROM
      table_name s
      )
```

**Actual behavior**

Crash:
```
token: {
    type: 'BETWEEN',
    raw: 'between',
    text: 'BETWEEN',
    start: 81,
    precedingWhitespace: ' '
  }
}
```

**Usage**
```
npx sql-formatter -l hive provoke.hql
```

Crash does not occur on 10.6.0
Exists in 11.0.2 and 12.0.0-beta.1

**Comment by nene** [comment]
Looks like there's a difference on what sort of nesting of operators is allowed.

The above query produces a syntax error in MySQL, but works in e.g. Postgres and SQLite.

But yeah, definitely a bug. Will fix.

**Comment by nene** [comment]
Unfortunately this is more complex problem than I had anticipated. It's very tricky to solve with our current parser architecture. I'm working on a longer-term solution, but that will take a whole lot of time to arrive.

For now the workaround is to wrap the between-expression inside parenthesis, like:

```sql
SELECT
  CASE
    WHEN (x BETWEEN 1 AND 10) THEN 1
  END
FROM
  tbl
```

**Comment by kbrownlees** [comment]
@nene is this sql a reproduction of the same bug or something new?

```sql
SELECT something
FROM the_table
WHERE
    field BETWEEN $1 AND $2
    AND field_2 = $3
```

Produces the error:

```
An Unexpected Error Occurred
Parse error: Ambiguous grammar
```

Also fixed by changing it to be:

```sql
SELECT something
FROM the_table
WHERE
    (field BETWEEN $1 AND $2)
    AND field_2 = $3
```

**Comment by nene** [comment]
@kbrownlees  This is duplicate of another bug: #534

Though, these two are somewhat related.

**Comment by QoVoQ** [comment]
Hi, @nene 
I noticed this line:  https://github.com/sql-formatter-org/sql-formatter/blob/58ac0fb4b2c802d4abdbc16e8888e417e894aa39/src/parser/createParser.ts#L42

May I known why ambiguous error should be thrown here?

When meet ambiguous grammar, can we just pick one of the parse results and carry on formatting process?

**Comment by nene** [comment]
When an ambiguity is encountered, it's a sign that the grammar itself is faulty, which means the bug is really inside the formatter and not in the specific input.

If we were to ignore the problem with ambiguity, the bug will likely show up later when formatting the result, but then it will be much harder to diagnose what went wrong.

It's an unfortunate state of this library that this ambiguity bug hasn't been fixed for many years now. But wiping it under the rug is a fix that's an invitation for many other bugs.

**Comment by gladykov** [comment]
Thank you very much sir!

**Comment by nene** [comment]
The fix for this is now released in 13.0.3
